File: openam-auth-trees/auth-nodes/src/main/java/org/forgerock/openam/auth/nodes/DebugNode.java
/*
 * The contents of this file are subject to the terms of the Common Development and
 * Distribution License (the License). You may not use this file except in compliance with the
 * License.
 *
 * You can obtain a copy of the License at legal/CDDLv1.0.txt. See the License for the
 * specific language governing permission and limitations under the License.
 *
 * When distributing Covered Software, include this CDDL Header Notice in each file and include
 * the License file at legal/CDDLv1.0.txt. If applicable, add the following below the CDDL
 * Header, with the fields enclosed by brackets [] replaced by your own identifying
 * information: "Portions copyright [year] [name of copyright owner]".
 *
 * Copyright 2025 ForgeRock AS.
 */
/*
 * Copyright 2022-2025 Ping Identity Corporation. All Rights Reserved
 *
 * This code is to be used exclusively in connection with Ping Identity
 * Corporation software or services. Ping Identity Corporation only offers
 * such software or services to legal entities who have entered into a
 * binding license agreement with Ping Identity Corporation.
 */
package org.forgerock.openam.auth.nodes;

import static org.forgerock.json.JsonValue.json;
import static org.forgerock.json.JsonValue.object;
import static org.forgerock.openam.audit.context.AuditRequestContext.getAuditRequestContext;
import static org.forgerock.openam.auth.node.api.Action.send;
import static org.forgerock.openam.auth.node.api.SharedStateConstants.REALM;

import java.util.ResourceBundle;
import java.util.Set;

import javax.security.auth.callback.TextOutputCallback;

import org.forgerock.json.JsonValue;
import org.forgerock.openam.annotations.sm.Attribute;
import org.forgerock.openam.audit.context.AuditRequestContext;
import org.forgerock.openam.auth.node.api.Action;
import org.forgerock.openam.auth.node.api.Node;
import org.forgerock.openam.auth.node.api.NodeProcessException;
import org.forgerock.openam.auth.node.api.NodeState;
import org.forgerock.openam.auth.node.api.SingleOutcomeNode;
import org.forgerock.openam.auth.node.api.TreeContext;
import org.forgerock.openam.auth.nodes.webauthn.ClientScriptUtilities;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.inject.Inject;
import com.google.inject.assistedinject.Assisted;
import com.sun.identity.authentication.callbacks.ScriptTextOutputCallback;

/**
 * A node which displays debug information about the current login tree. It
 * collects information from various sources such as the tree state (shared,
 * transient, secret), the identity objects universalId and the transaction id
 * which can be used as a reference in logs.
 */
@Node.Metadata(outcomeProvider = SingleOutcomeNode.OutcomeProvider.class,
      configClass = DebugNode.Config.class)
public class DebugNode extends SingleOutcomeNode {

    // This is used to grab the value of the transactionId
    private final AuditRequestContext auditRequestContext;
    // Utility which allows us to load a .js file as a string
    private final ClientScriptUtilities scriptUtils = new ClientScriptUtilities();
    // The javascript file we will later load in to a callback
    private static final String SCRIPT_PATH = "org/forgerock/openam/auth/nodes/DebugNode.js";
    // Bundle name that is used later in order to get locale translations
    private static final String BUNDLE_NAME = DebugNode.class.getName().replace(".", "/");
    // Logger instance which will allow us to add debug node logs to AM logs
    private final Logger logger = LoggerFactory.getLogger(DebugNode.class);
    // Stores the node's configuration attributes
    private final Config config;

    /**
     * Configuration for the debug node.
     */
    public interface Config {
        /**
        * Allow debug log popup to be turned on or off.
        *
        * @return whether popup is enabled.
        */
        @Attribute(order = 100)
        default boolean popupEnabled() {
            return true;
        }
    }

    /**
     * Create the node.
     * @param config The node config.
     */
    @Inject
    public DebugNode(@Assisted DebugNode.Config config) {
        this.config = config;
        this.auditRequestContext = getAuditRequestContext();
    }

    /**
     * Generates a json object of all the translations from the resource bundle
     * and returns it as a string, so it can be used by the loaded javascript file.
     *
     * @return stringified json
     */
    private String getTranslationsFromBundle(ResourceBundle bundle) {
        JsonValue translations = json(object());
        Set<String> keys = bundle.keySet();
        for (String key : keys) {
            translations.put(key, bundle.getString(key));
        }
        return translations.toString();
    }

    private String generateDebugScriptWithoutPopup() {
        return "window.frDebugNode.moveToNextNode();";
    }

    /**
     * Generates a new string of javascript which will add the latest debug log
     * to the page.
     *
     * @return string of javascript
     */
    private String generateDebugScriptWithPopup(String logs, TreeContext treeContext) {
        StringBuilder script = new StringBuilder();

        // Get the translated text for the node from the node bundle
        ResourceBundle bundle = treeContext.request.locales
                              .getBundleInPreferredLocale(BUNDLE_NAME, getClass().getClassLoader());
        String translations = getTranslationsFromBundle(bundle);

        // Create the popup window if it doesn't already exist
        script.append("window.frDebugNode.createPopup();");

        // Set the translations in JS so they can be used in the popup window
        script.append("window.frDebugNode.setTranslations('")
              .append(translations)
              .append("');");

        // Set the popup window title from translations
        script.append("window.frDebugNode.setPopupTitle('")
              .append(bundle.getString("popup.title"))
              .append("');");

        // Add the Realm name to the debug output
        JsonValue realm = treeContext.getStateFor(this).get(REALM);
        if (realm != null) {
            script.append("window.frDebugNode.setRealm('")
                  .append(realm.toString().replace("\"", ""))
                  .append("');");
        }

        // Add the new logs to the debug output
        script.append("window.frDebugNode.appendLog('")
              .append(logs)
              .append("');");

        // Auto submit the page to move to the next node in the tree
        script.append("window.frDebugNode.moveToNextNode();");

        return script.toString();
    }

    /**
     * Collects logs for the current state of the tree from treeContext.
     *
     * @return debug log
     */
    private String getLogs(TreeContext treeContext) {
        JsonValue log = json(object());

        // If the identity objects universalId exists, add it to the logs
        if (treeContext.universalId.isPresent()) {
            log.put("universalId", treeContext.universalId.get());
        }

        // Add the transactionId
        log.put("transactionId", this.auditRequestContext.getTransactionIdValue());

        // Get the state of the current node
        NodeState nodeState = treeContext.getStateFor(this);
        // Get all the nodes state keys
        Set<String> stateKeys = nodeState.keys();
        // Combine the keys with their values
        for (String key : stateKeys) {
            log.put(key, nodeState.get(key));
        }

        return log.toString();
    }

    @Override
    public Action process(TreeContext treeContext) throws NodeProcessException {
        // The process function is called before and after the node has been
        // shown. If the treeContext has callbacks this function has been
        // invoked at the end of the debug node lifecycle, and the tree should
        // proceed to the next step.
        if (treeContext.hasCallbacks()) {
            return goToNext().build();
        }

        String logs = getLogs(treeContext);
        // Add the latest logs to the debug logger
        logger.debug("DebugNode: {}", logs);

        String localScript;
        // If popup window is enabled, generate a script to update the popup
        // window and submit the page. If it is disabled, generate a script that
        // will submit the page only.
        if (config.popupEnabled()) {
            localScript = generateDebugScriptWithPopup(logs, treeContext);
        } else {
            localScript = generateDebugScriptWithoutPopup();
        }

        // Generate callbacks and send them to the page. When the script
        // callbacks are rendered, they will auto submit the page and move the
        // user to the next node. However, for a very brief period while the
        // scripts load, the login page will render and show a login button. A
        // TextOutputCallback has been used here to indicate to the user what is
        // happening instead of leaving the page blank.
        TextOutputCallback textCallback = new TextOutputCallback(TextOutputCallback.INFORMATION, "Debug Step");
        ScriptTextOutputCallback loadedScriptCallback = new ScriptTextOutputCallback(
            scriptUtils.getScriptAsString(SCRIPT_PATH));
        ScriptTextOutputCallback localScriptCallback = new ScriptTextOutputCallback(localScript);
        return send(loadedScriptCallback, localScriptCallback, textCallback).build();
    }
}

File: openam-auth-trees/auth-nodes/src/main/resources/org/forgerock/openam/auth/nodes/DebugNode.js
/*
 * The contents of this file are subject to the terms of the Common Development and
 * Distribution License (the License). You may not use this file except in compliance with the
 * License.
 *
 * You can obtain a copy of the License at legal/CDDLv1.0.txt. See the License for the
 * specific language governing permission and limitations under the License.
 *
 * When distributing Covered Software, include this CDDL Header Notice in each file and include
 * the License file at legal/CDDLv1.0.txt. If applicable, add the following below the CDDL
 * Header, with the fields enclosed by brackets [] replaced by your own identifying
 * information: "Portions copyright [year] [name of copyright owner]".
 *
 * Copyright 2022-2025 Ping Identity Corporation.
 */
// Ensure the frDebugNode object does not exist before we define the class
if (!window.frDebugNode) {
  /**
   * FrDebugNode Class for displaying tree debug logs in a popup window.
   * @class
   * @constructor
   * @public
   */
  class FrDebugNode {
    constructor() {
      /**
       * The name of the current login tree.
       * @type {String}
       * @public
       */
      this.tree = null;
      /**
       * The name of the current realm.
       * @type {String}
       * @public
       */
      this.realmName = null;
      /**
       * The popup window created to display logs in.
       * @type {Object}
       * @public
       */
      this.popup = null;
      /**
       * The title that will be given to the popup window.
       * @type {String}
       * @public
       */
      this.popupTitle = 'Tree Debugger';
      /**
       * Holds locale translated strings used throughout.
       * @type {String}
       * @public
       */
       this.tranlations = {
         'popup.title': 'Tree Debugger',
         'logs.initialStep': 'Initial Step',
         'logs.step': 'Step',
         'logs.realm': 'Realm'
       };
      /**
       * The array of logs being displayed. Each log contains the name
       * of the node, the date/time of the log and the log information (tree
       * state, transactionId, universalId).
       * @type {Array}
       * @public
       */
      this.logs = [];
      /**
       * Used to track the current node number and is displayed to the user
       * as the current step number.
       * @type {Number}
       * @public
       */
      this.nodeIndex = 0;
      /**
       * Some default styles that get loaded in to the popup.
       * @type {Number}
       * @public
       */
      this.css = `
        body { padding: 10px; }
        h1 { font-weight: 300; }
        h2 { color: rgb(42, 47, 50); margin: 0 0 40px 0; }
        h4 { margin-top: 0; margin-bottom: 0; display: inline-block; }
        p.fr-date { margin-bottom: 0; display: inline-block; }
        div.fr-log { margin-bottom: 40px; }
        pre {
          margin-top: 8px;
          background: #eee;
          line-height: 1.5;
          padding: 20px;
          border-radius: 4px;
          border: 1px solid #ddd;
          white-space: pre-wrap;
        }
      `;
    }

    /**
      * Creates the popup window that will display the logs, sets the window
      * title, some default titles and adds a placeholder element for the realm
      * name which will get set in the future.
      *
      */
    createPopup() {
      // Only create popup if it doesn't already exist
      if (this.popup === null) {
        this.popup = open('', 'debuggerWindow', `scrollbars=yes, width=${screen.availWidth / 2}, height=${screen.availHeight}`);

        if (!this.popup.document.querySelector('#fr-tree')) {
          this.popup.document.write(`
            <h2 id='fr-tree'></h2>
          `);
        }

        if (!this.popup.document.querySelector('#fr-realm')) {
          this.popup.document.write(`
            <h2 id='fr-realm'></h2>
          `);
        }


        const nativeStyles = document.querySelectorAll('link[rel=stylesheet]');
        if (nativeStyles.length) {
          for (let i = 0; i < nativeStyles.length; i++) {
            this.popup.document.head.append(nativeStyles[i].cloneNode(true));
          }
        }
        const localStyles = this.popup.document.createElement('style');
        localStyles.type = 'text/css';
        localStyles.appendChild(this.popup.document.createTextNode(this.css));
        this.popup.document.head.appendChild(localStyles);

        this.setPopupTitle(this.popupTitle);
      }
    }

    /**
      * Sets the locale translations used in the logs
      *
      * @param {String} tranlationString String of translations to be converted to json object
      */
    setTranslations(translationsString) {
      this.translations = JSON.parse(translationsString);
    };

    /**
      * Sets the html title of the popup window
      *
      * @param {String} title The title of the popup
      */
    setPopupTitle(title) {
      this.popup.document.title = title;
    }

    /**
      * Sets the current tree name in the popup window.
      *
      * @param {String} tree The name of the tree
      */
    setTree(tree) {
      this.tree = tree;
      this.popup.document.querySelector('#fr-tree').innerText = this.tree;
    }

    /**
      * Sets the name of the current realm in the popup window.
      *
      * @param {String} realm The name of the realm
      */
    setRealm(realm) {
      this.realm = realm;
      this.popup.document.querySelector('#fr-realm').innerText = `${this.translations['logs.realm']}: ${this.realm}`;
    }

    /**
      * Appends a new log to the list and displays it in the popup window. A log
      * contains the title of the current node (or index if it is not available),
      * a date/time of the log and log information such as the state of the tree.
      *
      * @param {String} realm The name of the realm
      */
    appendLog(log) {
      let date = new Date();
      // get a pretty printed version of the log
      const prettyLog = JSON.stringify(JSON.parse(log), null, 4);
      date = date.toLocaleDateString('en-US', { weekday: 'short', year: 'numeric', month: 'short', day: 'numeric', hour: 'numeric', minute: 'numeric', second: 'numeric' });
      const node = this.nodeIndex ? `${this.translations['logs.step']} ${this.nodeIndex}` : `${this.translations['logs.initialStep']}`;
      this.popup.document.write(`
        <div class='fr-log'>
          <div class='fr-log-title'>
            <h4>${node}</h4>
            <p class='fr-date text-primary'> - ${date}</p>
          </div>
          <pre>${prettyLog}</pre>
        </div>
      `);
      this.nodeIndex += 1;
    }

  /**
    * Every debug node renders a new login page which allows it to add new logs
    * to the popup window using javascript. Other than adding logs the page is
    * not very useful for the customer so this will auto submit the page to move
    * them along to the next node in the tree.
    */
    moveToNextNode() {
      document.querySelector('input[type=\"submit\"]').click();
    }
  }
  window.frDebugNode = new FrDebugNode();
}
